### DISCUSSION TITLE
Applying effects to a non arificially generated audio stream

### DISCUSSION THREAD

**Discussion by billy8119** [pschatzmann/arduino-audio-tools#319] [Q&A] [answered]
Microphone! I'm ultimately going to use an i2s stream, and I'm hoping portaudiostream is close enough for now, but I'm trying to add an effect to this stream.  Seems all the examples for effects use a generator of some kind or other, but not i2s as a sound SOURCE.  How do I use effects with a portaudiostream object (in this case microphone)
This didn't work:
#include "AudioTools.h"
#include "AudioLibs/PortAudioStream.h"

uint16_t sample_rate=22050;
uint8_t channels = 1;                                     // The stream will have 2 channels 
PortAudioStream aud; // On desktop we use 
AudioEffects effects(aud);
GeneratedSoundStream<int16_t> in(effects);

StreamCopy copier(aud,in); // copy in to out

// Arduino Setup
void setup(void) {
  Serial.begin(115200);
  
  AudioLogger::instance().begin(Serial, AudioLogger::Error);
  effects.addEffect(new Boost(4.0));
  // open output
  auto config = aud.defaultConfig();  
  config.sample_rate = sample_rate;
  config.channels = channels;
  config.bits_per_sample = sizeof(int16_t)*8;
  config.is_input=true;
  // aud.defaultConfig(RXTX_MODE);
  aud.begin(config);
}

// Arduino loop  
void loop() {
  if (aud)
    copier.copy();

....because there's no readSample method on the PortAudioStream object.  How SHOULD I read streaming audio one 16 bit sample at a time?
And if I'm reading a buffer at a time, do I need to rewrite the effects so they process one BUFFER at a time, instead of one WORD at a time?

**Comment by pschatzmann** [comment, marked as answer]
https://github.com/pschatzmann/arduino-audio-tools/wiki/Audio-Effects

**Reply by billy8119** [reply]
Yup.  Saw that.  (thank-you!) Same result - needs readSample.  In my case, I'm using PortAudioStream instead of AudioStream as I'm still prototyping on the desktop.
#include "AudioTools.h"
#include "AudioLibs/PortAudioStream.h"

uint16_t sample_rate=22050;
uint8_t channels = 1;                                     // The stream will have 2 channels 
PortAudioStream aud; // On desktop we use 
AudioEffects effects(aud);
AudioEffects<GeneratorFromStream<effect_t>> effects(aud,1, 1.0);
GeneratedSoundStream<int16_t> in(effects);

StreamCopy copier(aud,in); // copy in to out

// Arduino Setup
void setup(void) {
  Serial.begin(115200);
  
  AudioLogger::instance().begin(Serial, AudioLogger::Error);
  effects.addEffect(new Boost(4.0));
  // open output
  auto config = aud.defaultConfig();  
  config.sample_rate = sample_rate;
  config.channels = channels;
  config.bits_per_sample = sizeof(int16_t)*8;
  config.is_input=true;
  // aud.defaultConfig(RXTX_MODE);
  aud.begin(config);
}

// Arduino loop  
void loop() {
  if (aud)
    copier.copy();
}
Results in:
arduino_audio_tools-src/src/AudioEffects/AudioEffects.h: In instantiation of ‘audio_tools::effect_t audio_tools::AudioEffects<GeneratorT>::readSample() [with GeneratorT = audio_tools::PortAudioStream; audio_tools::effect_t = short int]’:
./arduino-audio-tools/examples/examples-desktop/generator/build/_deps/arduino_audio_tools-src/src/AudioEffects/AudioEffects.h:93:18:   required from here
./arduino-audio-tools/examples/examples-desktop/generator/build/_deps/arduino_audio_tools-src/src/AudioEffects/AudioEffects.h:96:40: error: ‘class audio_tools::PortAudioStream’ has no member named ‘readSample’
   96 |                 sample  = p_generator->readSample();
      |                           ~~~~~~~~~~~~~^~~~~~~~~~
make[2]: *** [CMakeFiles/generator.dir/build.make:76: CMakeFiles/generator.dir/generator.cpp.o] Error 1

**Comment by pschatzmann** [comment]
I think you forgot to delete the following original line which is causing the problem:
//AudioEffects effects(aud);
